mindmeld/components/custom_action.py

```diff
@@ -0,0 +1,202 @@
+        """Invoke the custom action with Request and Responder and return True if the action is
+        executed successfully, False otherwise. Upon successful execution, we update the Frame
+        and Directives of the Responder object.
+
+        Args:
+            request (Request)
+            responder (DialogueResponder)
+            async_mode (bool)
+
+        Returns:
+            (bool)
+        """
+        if not self.url:
+            raise CustomActionException(
+                "No URL is given for custom action {}.".format(self._name)
+            )
+
+        json_data = self.get_json_payload(request, responder)
+
+        try:
+            if async_mode:
+                # returning the coroutine to be awaited elsewhere
+                return self._process_async(json_data, responder)
+            else:
+                return self._process(json_data, responder)
+        except ConnectionError:
+            logger.error(
+                "Connection error trying to reach custom action server %s.", self.url
+            )
+            return False
+
+    async def invoke_async(self, request, responder):
+        """Asynchronously invoke the custom action with Request and Responder and return True if
+        the action is executed successfully, False otherwise. Upon successful execution, we update
+        the Frame and Directives of the Responder object.
+
+        Args:
+            request (Request)
+            responder (DialogueResponder)
+
+        Returns:
+            (bool)
+        """
+        return await self.invoke(request, responder, async_mode=True)
+
+    def _process(self, json_data, responder):
+        status_code, result_json = self.post(json_data)
+        return self._process_post_response(status_code, result_json, responder)
+
+    async def _process_async(self, json_data, responder):
+        status_code, result_json = await self.post_async(json_data)
+        return self._process_post_response(status_code, result_json, responder)
+
+    def _process_post_response(self, status_code, result_json, responder):
+        if status_code == 200:
+            for field in RESPONSE_FIELDS:
+                if field not in result_json:
+                    logger.warning(
+                        "`%s` not in the response of custom action %s",
+                        field,
+                        self._name,
+                    )
+            if self.merge:
+                responder.frame.update(result_json.get("frame", {}))
+                responder.directives.extend(result_json.get("directives", []))
+                responder.slots.update(result_json.get("slots", {}))
+            else:
+                responder.frame = result_json.get("frame", {})
+                responder.directives = result_json.get("directives", [])
+                responder.slots = result_json.get("slots", {})
+            responder.params = Params(**result_json.get("params", {}))
```

Comment:
Why is `responder.params = Params(**result_json.get("params", {}))` done when `self.merge == True`? I would also update the existing params if possible, no?

Reply:
Done